File: soul-server/src/soul_server/llm/adapters.py
"""
LLM Adapters - OpenAI/Anthropic 통합 인터페이스

각 LLM 프로바이더를 동일한 인터페이스로 호출하는 어댑터.
"""

import logging
from dataclasses import dataclass
from typing import Protocol

from soul_server.models.llm import LlmMessage

logger = logging.getLogger(__name__)


@dataclass
class LlmResult:
    """LLM 호출 결과"""
    content: str
    input_tokens: int
    output_tokens: int


class LlmAdapter(Protocol):
    """LLM 프로바이더 공통 인터페이스"""

    async def complete(
        self,
        model: str,
        messages: list[LlmMessage],
        max_tokens: int,
        temperature: float | None,
    ) -> LlmResult: ...


class OpenAIAdapter:
    """OpenAI API 어댑터"""

    def __init__(self, api_key: str) -> None:
        from openai import AsyncOpenAI

        self._client = AsyncOpenAI(api_key=api_key)

    async def complete(
        self,
        model: str,
        messages: list[LlmMessage],
        max_tokens: int,
        temperature: float | None,
    ) -> LlmResult:
        kwargs: dict = {
            "model": model,
            "messages": [{"role": m.role, "content": m.content} for m in messages],
            "max_tokens": max_tokens,
        }
        if temperature is not None:
            kwargs["temperature"] = temperature

        response = await self._client.chat.completions.create(**kwargs)

        if not response.choices:
            raise RuntimeError(
                f"OpenAI returned empty choices for model={model}"
            )
        content = response.choices[0].message.content or ""
        usage = response.usage

        return LlmResult(
            content=content,
            input_tokens=usage.prompt_tokens if usage else 0,
            output_tokens=usage.completion_tokens if usage else 0,
        )


class AnthropicAdapter:
    """Anthropic API 어댑터"""

    def __init__(self, api_key: str) -> None:
        from anthropic import AsyncAnthropic

        self._client = AsyncAnthropic(api_key=api_key)

    async def complete(
        self,
        model: str,
        messages: list[LlmMessage],
        max_tokens: int,
        temperature: float | None,
    ) -> LlmResult:
        # Anthropic은 system 메시지를 별도 파라미터로 전달
        system_parts: list[str] = []
        api_messages: list[dict] = []

        for m in messages:
            if m.role == "system":
                system_parts.append(m.content)
            else:
                api_messages.append({"role": m.role, "content": m.content})

        kwargs: dict = {
            "model": model,
            "messages": api_messages,
            "max_tokens": max_tokens,
        }
        if system_parts:
            kwargs["system"] = "\n\n".join(system_parts)
        if temperature is not None:
            kwargs["temperature"] = temperature

        response = await self._client.messages.create(**kwargs)

        # TextBlock에서 텍스트 추출
        content = ""
        for block in response.content:
            if hasattr(block, "text"):
                content += block.text

        return LlmResult(
            content=content,
            input_tokens=response.usage.input_tokens,
            output_tokens=response.usage.output_tokens,
        )

File: soul-server-ts/src/llm/adapters.ts
import type { LlmAdapter, LlmMessage, LlmResult } from "./types.js";

interface OpenAIChatCompletionResponse {
  choices?: Array<{
    message?: {
      content?: string | null;
    };
  }>;
  usage?: {
    prompt_tokens?: number;
    completion_tokens?: number;
  } | null;
  error?: {
    message?: string;
  };
}

interface AnthropicMessagesResponse {
  content?: Array<{
    type?: string;
    text?: string;
  }>;
  usage?: {
    input_tokens?: number;
    output_tokens?: number;
  };
  error?: {
    message?: string;
  };
}

export class OpenAIAdapter implements LlmAdapter {
  private readonly baseUrl: string;

  constructor(private readonly apiKey: string, baseUrl = "https://api.openai.com/v1") {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
  }

  async complete(params: {
    model: string;
    messages: LlmMessage[];
    maxTokens: number;
    temperature: number | null;
  }): Promise<LlmResult> {
    const body: Record<string, unknown> = {
      model: params.model,
      messages: params.messages.map((m) => ({ role: m.role, content: m.content })),
      max_completion_tokens: params.maxTokens,
    };
    if (params.temperature !== null) {
      body.temperature = params.temperature;
    }

    const response = await fetch(`${this.baseUrl}/chat/completions`, {
      method: "POST",
      headers: {
        authorization: `Bearer ${this.apiKey}`,
        "content-type": "application/json",
      },
      body: JSON.stringify(body),
    });
    const data = await readJson<OpenAIChatCompletionResponse>(response);
    if (!response.ok) {
      throw new Error(data.error?.message ?? `OpenAI request failed: ${response.status}`);
    }

    const content = data.choices?.[0]?.message?.content;
    if (typeof content !== "string") {
      throw new Error(`OpenAI returned empty choices for model=${params.model}`);
    }

    return {
      content,
      inputTokens: data.usage?.prompt_tokens ?? 0,
      outputTokens: data.usage?.completion_tokens ?? 0,
    };
  }
}

export class AnthropicAdapter implements LlmAdapter {
  private readonly baseUrl: string;

  constructor(private readonly apiKey: string, baseUrl = "https://api.anthropic.com") {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
  }

  async complete(params: {
    model: string;
    messages: LlmMessage[];
    maxTokens: number;
    temperature: number | null;
  }): Promise<LlmResult> {
    const systemParts: string[] = [];
    const apiMessages: Array<{ role: "user" | "assistant"; content: string }> = [];
    for (const message of params.messages) {
      if (message.role === "system") {
        systemParts.push(message.content);
      } else {
        apiMessages.push({ role: message.role, content: message.content });
      }
    }

    const body: Record<string, unknown> = {
      model: params.model,
      messages: apiMessages,
      max_tokens: params.maxTokens,
    };
    if (systemParts.length > 0) {
      body.system = systemParts.join("\n\n");
    }
    if (params.temperature !== null) {
      body.temperature = params.temperature;
    }

    const response = await fetch(`${this.baseUrl}/v1/messages`, {
      method: "POST",
      headers: {
        "x-api-key": this.apiKey,
        "anthropic-version": "2023-06-01",
        "content-type": "application/json",
      },
      body: JSON.stringify(body),
    });
    const data = await readJson<AnthropicMessagesResponse>(response);
    if (!response.ok) {
      throw new Error(data.error?.message ?? `Anthropic request failed: ${response.status}`);
    }

    const content = (data.content ?? [])
      .map((block) => (typeof block.text === "string" ? block.text : ""))
      .join("");

    return {
      content,
      inputTokens: data.usage?.input_tokens ?? 0,
      outputTokens: data.usage?.output_tokens ?? 0,
    };
  }
}

async function readJson<T>(response: Response): Promise<T> {
  const text = await response.text();
  if (!text) return {} as T;
  try {
    return JSON.parse(text) as T;
  } catch {
    throw new Error(`LLM provider returned non-JSON response: ${response.status}`);
  }
}
